syntax = "proto3";

import "trade_protocal.proto";
import "quote_protocal.proto";
import "riskapi_action_protocal.proto";

//namespace
package risk;

message error_info
{
    sint32          errorid=1;
    string          errormsg=2;
}

message rsp_head_info
{
    int32           requestid=1;
    error_info      info=2;
    bool            islast=3;
}

message rsp_risker_login
{
    int32           requestid=1;
    error_info      info=2;
}

message instrument_info
{
    string          contractcode=1;
    string          productcode=2;
    string          productkind=3;
    string          exchangecode=4;
    string          marketcode=5;
    string          createdate=6;
    string          opendate=7;
    string          expiredate=8;
    int32           multiplier=9;
    double          pricetick=10;
    string          contractname=11;
    string          delivbegindate=12;
    string          delivenddate=13;
    string          excontractname=14;
    string          exproductname=15;
    string          ismrgndisc=16;
    string          productclass=17;
    double          baseprice=18;
    string          optionstype=19;
    double          strikeprice=20;
    string          underlyinginstrid=21;
    string          strikemode=22;
    double          underlyingmultiple=23;
}

message rsp_instrument_field
{
    rsp_head_info   header=1;
    instrument_info field=2;
}

message rtn_quote_field
{
    quote_gateway_market_field field=1;
}

message rsp_position_field
{
    rsp_head_info header=1;
    trade_gateway_position_field field=2;
    double trypositionprofit=3;
}

message rtn_position_field
{
    trade_gateway_position_field field=1;
    double trypositionprofit=2;
}

message rsp_trade_field
{
    rsp_head_info             header=1;
    trade_gateway_trade_field field=2;
}

message rsp_order_field
{
    rsp_head_info             header=1;
    trade_gateway_order_field field=2;
}

message rtn_trade_field
{
    trade_gateway_trade_field field=1;
}

message rtn_order_field
{
    trade_gateway_order_field field=1;
}

message rsp_sync_account_field
{
    rsp_head_info                    header=1;
    trade_gateway_sync_account_field field=2;
}

message rtn_sync_account_field
{
    trade_gateway_sync_account_field field=1;
}

message shfe_margin_field
{
    double           pre_favorable=1;
    double           big_side_favorable=2;
}

message cffex_margin_field
{
    double           pre_favorable=1;
    double           speculation_favorable=2;
    double           hedging_favorable=3;
    double           arbitrage_favorable=4;
    double           market_favorable=5;
}

message dce_margin_field
{
    double           pre_favorable=1;
    double           sp_favorable=2;
    double           spc_favorable=3;
    double           locked_favorable=4;
}

message czce_margin_field
{
    double           pre_favorable=1;
    double           spd_favorable=2;
    double           ips_favorable=3;
    double           prt_favorable=4;
    double           std_favorable=5;
    double           stg_favorable=6;
    double           zprt_favorable=7;
    double           locked_favorable=8;
}

message ine_margin_field
{
    double           pre_favorable=1;
    double           big_side_favorable=2;
}

message delivery_margin_field
{
    double shfe_margin=1;
    double cffex_margin=2;
    double dce_margin=3;
    double czce_margin=4;
    double ine_margin=5;
}

message rtn_margin_field
{
    uint32 timestamp=1;
    string investor=2;
    shfe_margin_field shfe_margin=3;
    cffex_margin_field cffex_margin=4;
    dce_margin_field dce_margin=5;
    czce_margin_field czce_margin=6;
    ine_margin_field ine_margin=7;
    delivery_margin_field delivery_margin=8;
}

message rsp_margin_field
{
    rsp_head_info       header=1;
    rtn_margin_field    field=2;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
message rsp_risker_investor
{
    rsp_head_info header=1;
    string riskgroupid = 2;
    string investorid = 3;
}

message rsp_groups
{
    rsp_head_info header=1;
    repeated sys_group infos = 2;
}

message rsp_action_info
{
    int32 requestid=1;
    error_info info=2;
}

message rtn_change_risker
{
    //operatorid comment
    //1:insert group 2:remove group 
    //3:insert group investor 4:remove group investor 
    //5:insert risker investor 6:remove risker investor
    int32 operatorid = 1; 
    string userid = 2; //riskid
    sys_group group = 3;
    repeated string investor = 4;
}

///////////////////////////////////////////////////////////////////////////////
message rtn_option_position_risk
{
    //1:delta 2:gamma 3:theta 4:vega 5:rho 6:implied_volatility
    int32 type = 1; 
    int32 model = 2;
    string instrument = 3;
    double value = 4;
}

message rsp_whatif
{
    rsp_head_info header=1;
    int32 type = 2; //1:price 2:vol 3:rate
    int32 model = 3; //0:binomial 1:black 2:bs
    string instrument = 4;
    double range = 5;
    double option_value = 6;
}

message option_argv
{
    int32 model=1; //0:binomial 1:black 2:bs
    int32 price_type=2; //1:lastprice 2:presettlementprice 3:precloseprice 4:settlementprice
    double vol=3;
    double rate=4;
}

message rsp_option_argv
{
    rsp_head_info header=1;
    option_argv info=2;
}

message rtn_whatif
{
    whatif_argv field = 1;
    double option_value = 2;
}

